cpp/src/barretenberg/dsl/acir_format/acir_format.cpp

```diff
@@ -182,6 +198,21 @@ Composer create_circuit(const acir_format& constraint_system,
         create_block_constraints(composer, constraint);
     }
 
+    // Add recursion constraints
+    for (size_t i = 0; i < constraint_system.recursion_constraints.size(); ++i) {
+        auto& constraint = constraint_system.recursion_constraints[i];
+        create_recursion_constraints<false>(composer, constraint);
+
+        // make sure the verification key records the public input indices of the final recursion output
+        // (N.B. up to the ACIR description to make sure that the final output aggregation object wires are public
+        // inputs!)
+        if (i == constraint_system.recursion_constraints.size() - 1) {
+            std::vector<uint32_t> proof_output_witness_indices(constraint.output_aggregation_object.begin(),
```

Comment:
I added a method into `aggregation_state` that unifies these two steps: `void aggregation_state::assign_object_to_proof_outputs`

Would you mind updating to use it? I think it would allows us to remove the `set_recursive_proof` method and remove the complexity of having to assign `proof_output_witness_indices`

Reply:
It does look like we may still have to assign proof_output_witness_indices as I have to construct an aggregation_state in order to use `assign_object_to_proof_outputs`. It also looks like constructing the aggregation state will require fetching all the values of the `output_aggregation_state` from the composer to construct P0 and P1. I can still of course push an update to show how it would look but it does seem that there is some extra complexity to using `assign_object_to_proof_outputs` even if we can get rid of `set_recursive_proof`

Reply:
Oh that is a good point. One alternative is to pass in a `bool add_to_public_inputs` parameter into `create_recursion_constraint`. If `add_to_public_inputs == true` then we call `assign_object_to_proof_outputs` on `aggregation_state_ct result` (line 116)

Reply:
Unfortunately that alternative looks to also complicate how we currently construct our constraints. Calling `result.assign_object_proof_outputs` would mean we are now including the output aggregation object on the result aggregation state directly rather than it being a separate field on our RecursionConstraint. This change would require some larger changes in how we construct our recursion constraints and should be left as a separate optimization.

Reply:
sounds good. Might be a good idea to create a GH Issue to refactor the AggregationState interface to more easily handle this case + regular recursion w/o exposing internals like `proof_output_witness_indices`